src/Xamarin.Android.Build.Tasks/Tasks/AppendCustomMetadataToItemGroup.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.Android.Tasks {
+	public class AppendCustomMetadataToItemGroup : Task {
+		[Required]
+		public ITaskItem[] Inputs { get; set; }
+
+		[Required]
+		public ITaskItem[] MetaDataItems { get; set; }
+
+		[Output]
+		public ITaskItem[] Output { get; set; }
+
+		public override bool Execute ()
+		{
+			var output = new List<ITaskItem> ();
+			var metaData = new Dictionary<string, List<ITaskItem>> (StringComparer.InvariantCultureIgnoreCase);
+			foreach (ITaskItem item in MetaDataItems) {
+				if (!metaData.ContainsKey (item.ItemSpec))
+					metaData.Add (item.ItemSpec, new List<ITaskItem> ());
+				metaData[item.ItemSpec].Add (item);
+			}
+
+			foreach (var item in Inputs) {
+				var fn = Path.GetFileNameWithoutExtension (item.ItemSpec);
+				output.Add (item);
+				if (!metaData.ContainsKey (fn))
+					continue;
+				List<ITaskItem> metaDateList = metaData [fn];
+				if (metaDateList == null)
```

Comment:
This shouldn't be possible.  As per 4 statements earlier, `metaData` contains the key `fn`.  We thus *know* that `metaData[fn]` will return whatever we put in, and the only way `metaData[fn]` could be null is if we did `metaData.Add(key, null)`, which we don't.

Therefore, `metaDataList` cannot be null, and this `if` is unnecessary.

Reply:
We can probably use the TryGetValue as well:

```csharp
if (!metaData.TryGetValue(fn, out var metaDateList))
    continue;
```
then, we can reduce the accesses to a single operation.